Sources/HTTP/HTTPConnection.swift

```diff
@@ -0,0 +1,620 @@
+    return !hitParserError
+  }
+  
+  private let readBufferSize = 4096
+
+  internal func maybeReadMore() {
+    assert(isReadSideOpen, "starting, but reading-side is not open?")
+    guard !isReading     else { return } // still reading
+    guard isReadSideOpen else { return }
+    
+    isReading = true
+    channel.read(offset: 0, length: readBufferSize, queue: queue) {
+      // Note: This is retaining the connection! So if we stop, we also need to
+      //       stop this read-call! (it should error out if we close the read
+      //       side)
+      done, data, error in
+      
+      assert(self.isReading, "read callback called, but not reading anymore?")
+      guard self.isReadSideOpen else { return }
+      
+      if done {
+        self.isReading = false
+      }
+      
+      var hitParseError = false
+      
+      if error == 0, let data = data {
+        if done && data.isEmpty { // error = 0, empty data, + done means EOF
+          if self.isSuspended {
+            self.bufferedEOF = true
+          }
+          else {
+            hitParseError = !self.feedParser(nil)
+          }
+          self.isReadSideOpen = false
+        }
+        else if self.isSuspended {
+          if self.readBuffer == nil { self.readBuffer = data }
+          else { self.readBuffer!.append(data) }
+        }
+        else {
+          data.enumerateBytes { bptr, _, stop in
+            if !self.feedParser(bptr) {
+              hitParseError = true
+              stop = true
+            }
+          }
+        }
+      }
+      
+      if error != 0 || hitParseError {
+        self.abortOnError()
+      }
+      else if done && self.isReadSideOpen && !self.isSuspended {
+        // continue reading next block
+        self.maybeReadMore()
+      }
+    }
+  }
+  
+  /// A hard close, makes no sense to continue anything once we got a read
+  /// error.
+  func abortOnError() {
+    closeReadSide()
+    channel.close(flags: .stop)
+    bodyHandler = nil
+    for writer in writers {
+      writer.abort()
+    }
+    writers = []
+    
+    _connectionHasFinished()
+  }
+  
+  internal func serverWillStop() { // sent by server on server queue
+    queue.async { self._serverWillStop() }
+  }
+  private func _serverWillStop() {
+    // TODO: set some flag, teardown
+    // Note: the server object won't go away, we have a hard retain on it.
+  }
+  
+  /// Called by the writer if it is done and all queued writes have completed.
+  internal func _responseWriterIsDone(_ writer: HTTPResponseWriter) {
+    let isFirst = writers.first === writer
+    
+    guard isFirst else {
+      assert(writers.first == nil || writers.first === writer,
+             "writer done, but it is not the head writer? \(self) \(writer)")
+      return
+    }
+    
+    if let idx = writers.index(where: { $0 === writer }) {
+      // break retain cycle
+      writers.remove(at: idx)
+    }
+    else {
+      assert(false, "did not find writer which went done: \(self)")
```

Comment:
Can use `assertionFailure(message: String)`

Reply:
Don't care.

Reply:
Because that got a thumb dump, I guess I should clarify that what I meant is that I don't care whether or not we do this. I'm happy to do either :-) I don't care whether it is `assert(false)` or `assertionFailure`.